### PR TITLE
[compiler] Enum-driven visibility

### PR DESCRIPTION
**Pull request by witemple-msft**
This PR implements the design described in the Enum-Driven Visibility rework discussions.

There is still some work to be done:

- [x] Testing of new functionality (existing tests are almost all passing)
- [x] Diagnosing remaining failure w.r.t. Resource create operations in TypeSpec/REST
- [ ] Validate no breaking changes to Azure specifications.
- [x] Documentation review and updating existing visibility documentation to describe new behavior.

The design documents are located at the following gist: https://gist.github.com/witemple-msft/f122c1fce83e72d102290b5678930ac5

### CODE REVIEW

File: packages/compiler/src/lib/visibility.ts [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+  visibility: string | EnumValue,
+) => {
+  const keyProperties: ModelProperty[] = [];
+  entity.properties.forEach((prop: ModelProperty) => {
+    // Keep track of any key property without a visibility
+    // eslint-disable-next-line @typescript-eslint/no-deprecated
+    if (isKey(context.program, prop) && !getVisibility(context.program, prop)) {
+      keyProperties.push(prop);
+    }
+  });
```
Comment (timotheeguerin) [inline packages/compiler/src/lib/visibility.ts:87]: wish we could use the new `.map` on maps already :)

**Comment by markcowl** [comment]
Is there an uptake PR into typespec-azure?

**Review by timotheeguerin — APPROVED** [review]
Looks great overall, but yeah as mark said should double check any effect on typespec-azure + spec repo

**Comment by witemple-msft** [comment]
tsv passing: https://github.com/Azure/azure-rest-api-specs/pull/31469

**Comment by witemple-msft** [comment]
typespec-azure passing: https://github.com/Azure/typespec-azure/pull/1829

**Pull request merged**